Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,255 @@
+                                        CREDENTIAL_ENV_VAR));
+            }
+
+            // 2. Then try the well known file
+            try
+            {
+                credentialPath = GetWellKnownCredentialFilePath();
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return LoadCredentialFromFile(credentialPath);
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                // File is not present, eat the exception and move on to the next check.
+            }
+            catch (IOException e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
+                // but not limited to the System.IO or from the Newtonsoft.Json namespace
+
+                throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease rerun 'gcloud auth login' to regenerate credentials file.",
+                                        credentialPath,
+                                        e.Message));
+            }
+
+            // 3. Then try the compute engine, if not already checked         
+            if (!checkedComputeEngine)
+            {
+                Logger.Info("Checking whether the application is running on ComputeEngine.");
+                bool isRunningOnComputeEngine = ComputeCredential.IsRunningOnComputeEngine().Result;
+                checkedComputeEngine = true;
+                if (isRunningOnComputeEngine)
+                {
+                    Logger.Info("ComputeEngine check passed. Loading ComputeEngine Credentials.");
+                    return new ComputeCredential();
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Converts a stream to a credential object.
+        /// </summary>
+        internal static ICredential LoadFromStream(Stream stream)
+        {
+            ClientCredentialParameters clientCredentialParameters = ClientCredentialParameters.Load(stream);
+          
+            switch (clientCredentialParameters.CredentialType)
+            {
+                case ClientCredentialType.UserCredential:
+                    try
+                    {
+                        var initializer = new GoogleAuthorizationCodeFlow.Initializer
+                        {
+                            ClientSecrets = clientCredentialParameters.ClientSecrets
+                        };
+
+                        TokenResponse token = new TokenResponse()
+                        {
+                            RefreshToken = clientCredentialParameters.RefreshToken
+                        };
+
+                        var flow = new GoogleAuthorizationCodeFlow(initializer);
+                        return new UserCredential(flow, "ApplicationDefaultCredentials", token);
```

Comment:
It is not settable by the user. The scopes are set when the 'gcloud auth login' triggers the user consent flow. The scopes are determined by the various components that are integrated into the gcloud sdk.